### PR TITLE
contrib: update github login and fix orcid disconnect

### PR DESCRIPTION
**Pull request by hachreak**
* Fix github info retrieval.

* Fix orcid disconnection.

* Rely on github id instead of github primary email.

* Signup through security form.

* Prefill signup form.

* Force github registration through signup form.

* Makes signup form template configurable.

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

**Comment by hachreak** [comment]
ping @lnielsen

File: setup.py [hunk 1]
```diff
@@ -92,6 +92,7 @@
     'Flask-OAuthlib>=0.6.0,<0.7',  # quick fix for issue invenio#2158
     'Flask-Security>=1.7.5',
     'invenio-accounts>=1.0.0a6',
+    'invenio-db>=1.0.0a9',
```
Comment (lnielsen) [inline setup.py]: This one is already present in ``extra_requie``. You just have to install with pip install -e .[all,postgresql]

File: examples/github_app.py [hunk 2]
```diff
@@ -82,13 +84,14 @@
 
 from flask import Flask, redirect, url_for
 from flask_babelex import Babel
-from flask_cli import FlaskCLI
```
Comment (lnielsen) [inline examples/github_app.py]: This is still needed for ``Flask==0.10.x``

File: examples/requirements-github.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+invenio-oauthclient[github]
```
Comment (lnielsen) [inline examples/requirements-github.txt]: can these requirements go in setup?

File: tests/test_contrib_github.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    )
+
+
+def _get_state():
+    return serializer.dumps({'app': 'github', 'sid': _create_identifier(),
+                             'next': None, })
+
+
+def test_login(app):
+    """Test ORCID login."""
```
Comment (lnielsen) [inline tests/test_contrib_github.py]: ``/ORCID/GitHub/``

File: examples/requirements-github.txt (hunk 3 above)
Comment (hachreak) [inline examples/requirements-github.txt]: what do you mean? in the `setup.py`?

Comment (lnielsen) [inline examples/requirements-github.txt]: yes

File: invenio_oauthclient/contrib/github.py [hunk 5]
```diff
@@ -118,8 +118,8 @@ def account_info(remote, resp):
     ghuser = gh.me()
     email = ghuser.email
     if not email:
-        record = next(gh.iter_emails(1))
-        email = record['email'] if 'email' in record else None
+        record = next(gh.emails())
```
Comment (lnielsen) [inline invenio_oauthclient/contrib/github.py]:
```python
[{u'email': u'....', u'primary': False, u'verified': True},
 {u'email': u'....', u'primary': True, u'verified': True},]
```
Grab only the email if ``u'primary': True and u'verified': True``

Comment (lnielsen) [inline invenio_oauthclient/contrib/github.py]: - You should be able to sign in with a github user, logout, change github username, and login again.
- If no primary and verified email exists, then ask user to provide email address.

**Comment by lnielsen** [comment]
Another thing to test:

1) Create a local Invenio account with the *same username* as your github account, and **different email**.
2) Try to sign up using GitHub. This will create a conflict. Email doesn't exist in database, but username does. 

Currently I don't think the sign-up form can handle this. I.e. we should probably provide the entire signup form instead of only an email address

File: invenio_oauthclient/contrib/github.py [hunk 6]
```diff
@@ -146,3 +156,24 @@ def authorized(resp, remote):
...
+    external_ids = [i.id for i in current_user.external_identifiers
+                    if i.method == external_method]
+
+    if external_ids:
+        oauth_unlink_external_id(dict(id=external_ids[0],
+                                      method=external_method))
+    if remote_account:
+        with db.session.begin_nested():
+            remote_account.delete()
+
```
Comment (jirikuncar) [inline invenio_oauthclient/contrib/github.py]: `db.session.commit` is usually needed before redirect

File: tests/conftest.py [hunk 7]
```diff
@@ -61,8 +62,13 @@ def base_app(request):
         LOGIN_DISABLED=False,
         CACHE_TYPE='simple',
         OAUTHCLIENT_REMOTE_APPS=dict(
-            orcid=ORCHID_REMOTE_APP,
             cern=CERN_REMOTE_APP,
+            orcid=ORCID_REMOTE_APP,
+            github=GITHUB_REMOTE_APP
```
Comment (jirikuncar) [inline tests/conftest.py]: please add `,` at the end of line

File: tests/test_app.py [hunk 8]
```diff
@@ -29,10 +29,9 @@
 from flask_oauthlib.client import OAuth as FlaskOAuth
 from flask_oauthlib.client import OAuthRemoteApp
 from invenio_db import InvenioDB, db
-from sqlalchemy_utils.functions import create_database, database_exists
-
 from invenio_oauthclient import InvenioOAuthClient
 from invenio_oauthclient.contrib.orcid import REMOTE_APP
+from sqlalchemy_utils.functions import create_database, database_exists
```
Comment (jirikuncar) [inline tests/test_app.py]: IMHO this should be above - check if `sqlalchemy_utils` is in known third party packages. Have you installed it in dev mode before?

File: tests/test_contrib_github.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015, 2016 CERN.
```
Comment (jirikuncar) [inline tests/test_contrib_github.py]: Why 2015? Did you copy it from somewhere?

File: tests/test_contrib_orcid.py [hunk 10]
```diff
@@ -140,6 +139,9 @@ def test_authorized_signup(app, example_orcid, orcid_bio):
             method='orcid', id_user=user.id,
             id=example_data['orcid']
         ).count()
+        assert len(RemoteAccount.query.filter_by(id=user.id).all()) == 0
```
Comment (jirikuncar) [inline tests/test_contrib_orcid.py]: is there any reason why you can not use `count()` instead of `len(...all())`?

File: invenio_oauthclient/contrib/github.py (hunk 6 above)
Comment (hachreak) [inline invenio_oauthclient/contrib/github.py]: I added the commit here on oauthclient `disconnect()` (because also orcid needed commit): https://github.com/inveniosoftware/invenio-oauthclient/pull/54/commits/7c22ee2f526b3a3360ba6edbe42c16f221fea49e#diff-f2d8bbac55a20ee4eb23a17827b21543R151
WDYT?

File: tests/test_app.py (hunk 8 above)
Comment (hachreak) [inline tests/test_app.py]: I tried to add as `known_third_party` but I have the same result from `isort`.
Do you have any other suggestion to fix it?

Comment (jirikuncar) [inline tests/test_app.py]: First try to completely uninstall `sqlalchemy_utils`. You might need to run `pip uninstall sqlalchemy-utils` multiple times. Then you can try to install it from pypi and run the isort again.

File: invenio_oauthclient/templates/invenio_oauthclient/signup.html [hunk 11]
```diff
@@ -22,31 +22,31 @@
...
+        <h3> {{ _('Sign-up with %(title)s!', title=app_title) }}</h3>
+        <p>{{ _('Fill in your details to complete your registration. You only have to do this once.') }}</p>
+        <hr />
+      {% endblock %}
+      {%- block signup_form %}
+      <form name="edit" method="post" role="form">
+        {%- for field in form %}
+        {{ render_field(field) }}
+        {%- endfor %}
+        <!-- <div class="form-actions"> -->
```
Comment (jirikuncar) [inline invenio_oauthclient/templates/invenio_oauthclient/signup.html]: @hachreak why is it commented out?

File: invenio_oauthclient/utils.py [hunk 12]
```diff
@@ -91,7 +91,7 @@ def oauth_authenticate(client_id, user, require_existing_link=False,
 
 def oauth_register(account_info, form_data=None):
     """Register user if possible."""
-    email = account_info.get("email")
+    email = account_info.get('user', {}).get("email")
```
Comment (jirikuncar) [inline invenio_oauthclient/utils.py]: can you please replace `"` with `'` so it's consistent?

File: setup.cfg [hunk 13]
```diff
@@ -31,6 +28,9 @@ all_files = 1
 [bdist_wheel]
 universal = 1
 
+[aliases]
```
Comment (jirikuncar) [inline setup.cfg]: Please revert this change so the sections are alphabetically sorted. Thank you!

File: setup.py [hunk 14]
```diff
@@ -57,9 +56,6 @@
     'github': [
         'github3.py>=1.0.0a4',
     ],
-    'orcid': [
```
Comment (jirikuncar) [inline setup.py]: please keep the `'orcid'` section even when it's empty for compatibility reasons.

File: tests/conftest.py [hunk 15]
```diff
@@ -37,17 +37,21 @@
...
-    drop_database
-
 from invenio_oauthclient import InvenioOAuthClient
-from invenio_oauthclient.contrib.orcid import REMOTE_APP as ORCHID_REMOTE_APP
 from invenio_oauthclient.contrib.cern import REMOTE_APP as CERN_REMOTE_APP
+from invenio_oauthclient.contrib.github import REMOTE_APP as GITHUB_REMOTE_APP
+from invenio_oauthclient.contrib.orcid import REMOTE_APP as ORCID_REMOTE_APP
 from invenio_oauthclient.views.client import blueprint as blueprint_client
 from invenio_oauthclient.views.settings import blueprint as blueprint_settings
+from invenio_userprofiles import InvenioUserProfiles
```
Comment (jirikuncar) [inline tests/conftest.py]: you should add `invenio_userprofiles` to the `known_thirdparty` section

File: tests/helpers.py [hunk 16]
```diff
@@ -21,9 +21,8 @@
 
 from flask import jsonify
 from flask_login import _create_identifier
-from mock import MagicMock
-
 from invenio_oauthclient.views.client import serializer
+from mock import MagicMock
```
Comment (jirikuncar) [inline tests/helpers.py]: revert this change

File: tests/test_admin.py [hunk 17]
```diff
@@ -22,11 +22,10 @@
 from flask.ext.admin import Admin
 from invenio_admin import InvenioAdmin
 from invenio_db import db
-from werkzeug.local import LocalProxy
-
 from invenio_oauthclient import InvenioOAuthClient
 from invenio_oauthclient.admin import RemoteAccountView, RemoteTokenView, \
     remote_account_adminview, remote_token_adminview
+from werkzeug.local import LocalProxy
```
Comment (jirikuncar) [inline tests/test_admin.py]: revert this change

File: invenio_oauthclient/templates/invenio_oauthclient/signup.html (hunk 11 above)
Comment (hachreak) [inline invenio_oauthclient/templates/invenio_oauthclient/signup.html]: sorry, I forgot to remove.
I removed because I'm using the signup form that already have a "register" button.

File: tests/test_contrib_cern.py [hunk 18]
```diff
@@ -21,15 +21,14 @@
 
 from __future__ import absolute_import
 
-from flask import session, url_for
-from six.moves.urllib_parse import parse_qs, urlparse
-
 import invenio_oauthclient.contrib.cern as cern
-from invenio_oauthclient.models import RemoteToken
+from flask import session, url_for
```
Comment (jirikuncar) [inline tests/test_contrib_cern.py]: please make sure that the sections are restored as they were

File: setup.py (hunk 14 above)
Comment (hachreak) [inline setup.py]: done

File: tests/conftest.py (hunk 15 above)
Comment (hachreak) [inline tests/conftest.py]: done

File: tests/helpers.py (hunk 16 above)
Comment (hachreak) [inline tests/helpers.py]: do I need to configure in some way isort?

File: tests/test_contrib_cern.py (hunk 18 above)
Comment (hachreak) [inline tests/test_contrib_cern.py]: the sections? Can you explain me better what do you mean? thanks.

Comment (jirikuncar) [inline tests/test_contrib_cern.py]: Import sections - in the context of other comments.

File: setup.py [hunk 19]
```diff
@@ -89,15 +86,17 @@
...
-    'invenio-accounts>=1.0.0a6',
+    'invenio-accounts>=1.0.0a8',
+    'invenio-mail>=1.0.0a2',
     # FIXME
     #  'invenio-upgrader>=0.1.0',
     # FIXME new oauthlib release after 0.7.2 has some compatible problems with
     # the used Flask-Oauthlib version.
     'oauthlib==0.7.2',
-    'six>=1.7.2',
+    'six>=1.9',
```
Comment (jirikuncar) [inline setup.py]: is there any problem with older version of the `six` library?

Comment (hachreak) [inline setup.py]: I had a version conflict on lowest requirements:
```python 
Using /src/code/.eggs/Flask_BabelEx-0.9.3-py2.7.egg
Traceback (most recent call last):
  File "setup.py", line 166, in <module>
    'Development Status :: 1 - Planning',
  File "/usr/local/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/local/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/local/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "build/bdist.linux-x86_64/egg/ptr.py", line 73, in run
  File "/usr/local/lib/python2.7/site-packages/setuptools/dist.py", line 313, in fetch_build_eggs
    replace_conflicting=True,
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 826, in resolve
    dist = best[req.key] = env.best_match(req, ws, installer)
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1085, in best_match
    dist = working_set.find(req)
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 695, in find
    raise VersionConflict(dist, req)
pkg_resources.VersionConflict: (six 1.7.2 (/usr/local/lib/python2.7/site-packages), Requirement.parse('six>=1.9'))
```

File: tests/helpers.py (hunk 16 above)
Comment (jirikuncar) [inline tests/helpers.py]: It depends how did you install the package dependencies.

For more details please see http://isort.readthedocs.io/en/latest/#configuring-isort

File: invenio_oauthclient/contrib/github.py [hunk 20]
```diff
@@ -112,22 +117,36 @@
...
-        email = record['email'] if 'email' in record else None
-    # FIXME get user informations accordly with invenio-userprofiles
-    return dict(email=email, external_id=ghuser.id,
-                external_method='github')
+    me = gh.me()
+    return dict(
+        user=dict(
+            email=_extract_email(gh),
+            profile=dict(
+                username=me.name,
```
Comment (lnielsen) [inline invenio_oauthclient/contrib/github.py]: should this not be ``me.login``?

File: invenio_oauthclient/handlers.py [hunk 21]
```diff
@@ -234,20 +235,18 @@ def authorized_signup_handler(resp, remote, *args, **kwargs):
 
         if user is None:
             # Auto sign-up if user not found
-            user = oauth_register(account_info)
```
Comment (lnielsen) [inline invenio_oauthclient/handlers.py]: Shouldn't the ``if`` statement stay?

File: invenio_oauthclient/handlers.py [hunk 22]
```diff
@@ -234,20 +235,18 @@ def authorized_signup_handler(resp, remote, *args, **kwargs):
...
-            if user is None:
```
Comment (hachreak) [inline invenio_oauthclient/handlers.py]: @lnielsen do you mean this `if`?
I needed to remove to force to pass from `signup` form in any case.

File: invenio_oauthclient/handlers.py [hunk 23]
```diff
@@ -373,6 +376,10 @@ def signup_handler(remote, *args, **kwargs):
         else:
             return redirect('/')
 
+    # prefill form
+    account_info = session.get(session_prefix + '_account_info')
+    form = prefill_form(form, account_info['user'])
+
     return render_template(
```
Comment (lnielsen) [inline invenio_oauthclient/handlers.py]: Can we change ``'invenio_oauthclient/signup.html'`` into a configuration variable, so it's possible to override the template in an overlay.

File: tests/test_contrib_github.py [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+            client.get(url_for("invenio_oauthclient.login",
+                               remote_app='github'))
+
+            # Mock access token request
+            mock_response(app.extensions['oauthlib.client'], 'github',
+                          example_github)
+
+            # User then goes to "Linked accounts" and clicks "Connect"
+            resp = client.get(
+                url_for("invenio_oauthclient.login", remote_app='github',
```
Comment (jirikuncar) [inline tests/test_contrib_github.py]: `s/"/'/`

File: invenio_oauthclient/handlers.py (hunk 22 above)
Comment (jirikuncar) [inline invenio_oauthclient/handlers.py]: @lnielsen it looks correct since @hachreak added https://github.com/inveniosoftware/invenio-oauthclient/pull/54/files#diff-63911819994c3bdd7c57acc33a482540R380

Comment (lnielsen) [inline invenio_oauthclient/handlers.py]: @jirikuncar I don't think so, since then it will always display the sign-up form, instead of just register a user if all information is available?

File: invenio_oauthclient/handlers.py (hunk 23 above)
Comment (hachreak) [inline invenio_oauthclient/handlers.py]: done

**Comment by hachreak** [comment]
ping @lnielsen 
(see the second commit for latest changes)

File: invenio_oauthclient/handlers.py (hunk 22 above)
Comment (jirikuncar) [inline invenio_oauthclient/handlers.py]: Hmm, that's questionable. I would like to see the registration form with pre-populated values, but I can image that in some use-cases it might be confusing for the end-user.

File: invenio_oauthclient/templates/invenio_oauthclient/signup.html [hunk 25]
```diff
@@ -22,31 +22,27 @@
...
-          <div class="form-actions">
-            <a href="/" class="btn btn-default"><i class="fa fa-times"></i> Cancel</a>
-            <button type="submit" name="continue" class="btn btn-warning"/><i class="fa fa-edit"></i> {{ _("Sign-up") }}</button>
-          </div>
-        </form>
+      {%- block signup_header %}
+      {%- block signup_app_icon %}
+      {% if app_icon -%}
+        <span class="pull-right"><i class="{{app_icon}} fa-5x"></i></span>
+        {%- endif %}
```
Comment (jirikuncar) [inline invenio_oauthclient/templates/invenio_oauthclient/signup.html]: can you check the indentation here?

File: invenio_oauthclient/utils.py [hunk 26]
```diff
@@ -89,23 +91,22 @@ def oauth_authenticate(client_id, user, require_existing_link=False,
...
-
-    if email:
-        data = dict(**form_data) if form_data else {}
-        data['password'] = ''
-        data['email'] = email
+    data = deepcopy(form.data)
+    form = disable_csrf_form(form)
+
+    if form.validate():
+        if 'password' not in data or not data['password']:
```
Comment (jirikuncar) [inline invenio_oauthclient/utils.py]: Do you ever want to allow user authentication with password for user registered via OAuth?

File: invenio_oauthclient/handlers.py [hunk 27]
```diff
@@ -373,8 +387,12 @@ def signup_handler(remote, *args, **kwargs):
         else:
             return redirect('/')
 
+    # prefill form
+    account_info = session.get(session_prefix + '_account_info')
+    form = prefill_form(form, account_info['user'])
```
Comment (jirikuncar) [inline invenio_oauthclient/handlers.py]: we should be careful not to override the form data on POST.

File: invenio_oauthclient/handlers.py [hunk 28]
```diff
@@ -315,18 +324,23 @@ def signup_handler(remote, *args, **kwargs):
...
-    if not session.get(session_prefix + '_autoregister',
-                       False):
+    if not session.get(session_prefix + '_autoregister', False):
         return redirect(url_for('.login', remote_app=remote.name))
 
-    form = EmailSignUpForm(request.form)
+    security_ext = current_app.extensions['security']
+    form = security_ext.confirm_register_form(request.form)
 
     if form.validate_on_submit():
```
Comment (jirikuncar) [inline invenio_oauthclient/handlers.py]: we might need to split validation and check for request method.

File: invenio_oauthclient/handlers.py [hunk 29]
```diff
@@ -234,9 +235,17 @@ def authorized_signup_handler(resp, remote, *args, **kwargs):
 
         if user is None:
             # Auto sign-up if user not found
-            user = oauth_register(account_info)
+            security_ext = current_app.extensions['security']
```
Comment (jirikuncar) [inline invenio_oauthclient/handlers.py]: what about creating proxy for `confirm_register_form`

File: invenio_oauthclient/handlers.py (hunk 21 above)
Comment (hachreak) [inline invenio_oauthclient/handlers.py]: done

File: invenio_oauthclient/utils.py (hunk 26 above)
Comment (hachreak) [inline invenio_oauthclient/utils.py]: ping @lnielsen

Comment (switowski) [inline invenio_oauthclient/utils.py]: Just a small code-style advice:
`if 'password' not in data or not data['password']:` == `if not data.get('password')`
:)

File: invenio_oauthclient/handlers.py (hunk 29 above)
Comment (hachreak) [inline invenio_oauthclient/handlers.py]: what do you mean exactly? 
should I create a function in `utils.py` to retrieve the `confirm_register_form`?
like:
```python
def current_confirm_register_form():
    security_ext = current_app.extensions['security']
    return security_ext.confirm_register_form
```

Comment (jirikuncar) [inline invenio_oauthclient/handlers.py]: or `current_confirm_register_form = LocalProxy(lambda: current_app.extensions['security'].confirm_register_form)`

**Comment by lnielsen** [comment]
@hachreak Can you take a look why the tests are not passing.

**Comment by hachreak** [comment]
ok, I'll take a look

**Comment by hachreak** [comment]
ping @lnielsen
